shadow-utils-4.1.4.2/man/logoutd.8
'\" t
.\"     Title: logoutd
.\"    Author: [FIXME: author] [see http://docbook.sf.net/el/author]
.\" Generator: DocBook XSL Stylesheets v1.75.1 <http://docbook.sf.net/>
.\"      Date: 07/24/2009
.\"    Manual: System Management Commands
.\"    Source: System Management Commands
.\"  Language: English
.\"
.TH "LOGOUTD" "8" "07/24/2009" "System Management Commands" "System Management Commands"
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
logoutd \- Enforce login time restrictions
.SH "SYNOPSIS"
.HP \w'\fBlogoutd\fR\ 'u
\fBlogoutd\fR
.SH "DESCRIPTION"
.PP

\fBlogoutd\fR
enforces the login time and port restrictions specified in
/etc/porttime\&.
\fBlogoutd\fR
should be started from
/etc/rc\&. The
/var/run/utmp
file is scanned periodically and each user name is checked to see if the named user is permitted on the named port at the current time\&. Any login session which is violating the restrictions in
/etc/porttime
is terminated\&.
.SH "FILES"
.PP
/etc/porttime
.RS 4
File containing port access\&.
.RE
.PP
/var/run/utmp
.RS 4
List of current login sessions\&.
.RE


gmime-38b35d8/ChangeLog
2015-01-23  Sunghyun Kwon     <sh0701.kwon@samsung.com>

    * gmime/gmime-gpg-context.c (gpg_ctx_set_mode): Added the condition
    (GPG_CTX_MODE_SIGN_ENCRYPT) 

2013-02-19  Jeffrey Stedfast  <fejj@gnome.org>

	* README: Bumped version

	* configure.ac: Bumped version to 2.6.15

2013-02-19  Jeffrey Stedfast  <fejj@gnome.org>

	* gmime/gmime-filter-html.c (html_convert): Fixed to convert the
	last line of input even if it doesn't end with a \n.

2013-02-18  Jeffrey Stedfast  <fejj@gnome.org>

	* README: Bumped version

	* configure.ac: Bumped version to 2.6.14

2013-02-18  Jeffrey Stedfast  <fejj@gnome.org>

	* util/url-scanner.c (url_addrspec_start): Fixed the logic to
	actually scan for the beginning of the mailbox.

2012-12-09  Jeffrey Stedfast  <fejj@gnome.org>

	* README: Bumped version

	* configure.ac: Bumped version to 2.6.13

2012-11-26  Jeffrey Stedfast  <fejj@gnome.org>

	* gmime/gmime-header.c (g_mime_header_iter_first): Make sure the
	list isn't empty.
	(g_mime_header_iter_last): Same.

2012-11-23  Jeffrey Stedfast  <fejj@gnome.org>

	* gmime/gmime-header.c (g_mime_header_list_prepend): Prepend the
	header to the list, don't *append* it. Oops.
	(g_mime_header_list_clear): Don't forget to reset the linked list
	so that it doesn't contained free'd nodes.

2012-11-02  Jeffrey Stedfast  <fejj@gnome.org>

	* README: Bumped version

	* configure.ac: Bumped version to 2.6.12

2012-10-27  Jeffrey Stedfast  <fejj@gnome.org>

	* gmime/gmime-stream-filter.c (stream_read): Initialize presize to
	READ_PAD instead of READ_SIZE to prevent potential buffer
	underruns.

2012-10-20  Jeffrey Stedfast  <fejj@gnome.org>

	* gmime/gmime-filter-gzip.c (gunzip_filter): Apparently
	Z_BUF_ERROR isn't actually an error unless there is input
	available. Thanks to Pavlo <devgs@ukr.net> for this patch.

2012-10-19  Jeffrey Stedfast  <fejj@gnome.org>

	* gmime/gmime-parser.c (parser_push_boundary): Don't use strcmp to
	check if the boundary is an mbox boundary, use a direct pointer
	comparison so that we can handle Content-Type boundary parameters
	with a value of "From "

2012-10-19  Jeffrey Stedfast  <jeff@xamarin.com>

	* gmime/gmime-parser.c (parser_step_headers): Don't assume that
	the first non-header line we find indicates the start of the
	content unless we've already gotten some basic headers.

	Fix for bug #686223.

2012-10-18  Jeffrey Stedfast  <jeff@xamarin.com>

	* README: Bumped version

	* configure.ac: Bumped version to 2.6.11

2012-09-12  Jeffrey Stedfast  <jeff@xamarin.com>

	* gmime/gmime-gpg-context.c (gpg_ctx_parse_status): Unref the
	certificate after adding it to the list (adding it to the list
	refs it).

	* gmime/gmime-multipart-encrypted.c
	(g_mime_multipart_encrypted_decrypt): If the 'result' out param is
	null, unref the GMimeDecryptResult to avoid leaking.

2012-08-13  Jeffrey Stedfast  <jeff@xamarin.com>

	* gmime/gmime-utils.c (tokenize_rfc2047_phrase): Don't forget to
	initialize 'ascii' to TRUE.

2012-08-05  Jeffrey Stedfast  <jeff@xamarin.com>

	* gmime/gmime-stream-pipe.c (stream_flush): Pipes and sockets
	cannot be fsync()'d, so don't even try!

2012-08-05  Jeffrey Stedfast  <jeff@xamarin.com>

	* gmime/gmime-utils.c (rfc2047_token_new_encoded_word): Make sure
	to properly handle broken encoded-word tokens in the form:
	=?charset?q?=.

	Fixes the g_malloc abort() in bug #679388

2012-08-05  Jeffrey Stedfast  <jeff@xamarin.com>

	* gmime/gmime-multipart-signed.c (check_protocol_supported): Oops,
	need to compare against xsupported in the second string compare.

	Fixes bug #674032

2012-06-13  Jeffrey Stedfast  <fejj@gnome.org>

	* README: Bumped version

	* configure.ac: Bumped version to 2.6.10

2012-06-13  Jeffrey Stedfast  <fejj@gnome.org>

	* gmime/Makefile.am: Always build gmime-gpg-context.[c,h] and
	gmime-pkcs7-context.[c,h]

	* gmime/gmime-gpg-context.c: Armor internals with #ifdef
	ENABLE_CRYPTOGRAPHY so that if crypto is disabled, only the
	skeletal API is built.

	* gmime/gmime.h: Include gmime-pkcs7-context.h

	Fixes bug #676692.

2012-06-13  Jeffrey Stedfast  <fejj@gnome.org>

	* gmime/gmime-utils.c (rfc2047_encode_get_rfc822_words): Use
	is_blank() instead of is_lwsp() so that we only treat space and
	tab as word delimiters (we want to rfc2047 encode \r's and
	\n's). Check for ctrl characters when deciding what the type and
	encoding of the word is.

2012-06-13  Jeffrey Stedfast  <fejj@gnome.org>

	* gmime/gmime-utils.c (rfc2047_encode_get_rfc822_words): When
	breaking up really long atom tokens, change the word type to
	RFC2047 so that we encode them, thus preventing forced header
	folding which cannot be unmunged.
	(header_fold): Fixed for cases where the first word in a header is
	68 chars. (Note: this fix is largely unnecessary due to the above fix)

2012-06-13  Jeffrey Stedfast  <fejj@gnome.org>

	* gmime/gmime-message.c: When setting a header, make sure to
	always clear the header list cache of the toplevel mime_part as
	well.

2012-06-13  Jeffrey Stedfast  <fejj@gnome.org>

	* gmime/gmime-param.c (rfc2184_param_new): Only decode the charset
	and lang if the parameter is actually encoded. Thanks to Peter
	Bloomfield for this patch.

2012-06-13  Jeffrey Stedfast  <fejj@gnome.org>

	* gmime/gmime-gpg-context.c (gpg_ctx_parse_status): Keep track of
	DECRYPTION_OKAY status.
	(gpg_decrypt): Only return NULL if decryption failed (e.g. we
	didn't get a DECRYPTION_OKAY status).

	Fixes bug #677088.

2012-06-13  Jeffrey Stedfast  <fejj@gnome.org>

	* docs/reference/building.sgml: Updated docs.

	Fixes bug #676341

2012-04-22  Jeffrey Stedfast  <fejj@gnome.org>

	* README: Bumped version

	* configure.ac: Bumped version to 2.6.9

2012-04-22  Jeffrey Stedfast  <fejj@gnome.org>

	* gmime/gmime-multipart-signed.c (g_mime_multipart_signed_verify):
	Match protocols like application/x-pkcs7-signature, too.

	Fixes bug #674032.

2012-04-19  Jeffrey Stedfast  <fejj@gnome.org>

	* README: Bumped version

	* configure.ac: Bumped version to 2.6.8


shadow-utils-4.1.4.2/src/logoutd.c
/*
 * Copyright (c) 1991 - 1993, Julianne Frances Haugh
 * Copyright (c) 1996 - 2000, Marek Michałkiewicz
 * Copyright (c) 2001 - 2006, Tomasz Kłoczko
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. The name of the copyright holders or contributors may not be used to
 *    endorse or promote products derived from this software without
 *    specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
 * PARTICULAR PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT
 * HOLDERS OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#include <config.h>

#ident "$Id: logoutd.c 2849 2009-04-30 21:08:49Z nekral-guest $"

#include <fcntl.h>
#include <signal.h>
#include <stdio.h>
#include <sys/stat.h>
#include <sys/types.h>
#include "defines.h"
#include "prototypes.h"
/*
 * Global variables
 */
char *Prog;

#ifndef DEFAULT_HUP_MESG
#define DEFAULT_HUP_MESG _("login time exceeded\n\n")
#endif

#ifndef HUP_MESG_FILE
#define HUP_MESG_FILE "/etc/logoutd.mesg"
#endif

/* local function prototypes */
#ifdef USE_UTMPX
static int check_login (const struct utmpx *ut);
#else				/* !USE_UTMPX */
static int check_login (const struct utmp *ut);
#endif				/* !USE_UTMPX */
static void send_mesg_to_tty (int tty_fd);

/*
 * check_login - check if user (struct utmpx/utmp) allowed to stay logged in
 */
#ifdef USE_UTMPX
static int check_login (const struct utmpx *ut)
#else				/* !USE_UTMPX */
static int check_login (const struct utmp *ut)
#endif				/* !USE_UTMPX */
{
	char user[sizeof (ut->ut_user) + 1];
	time_t now;

	/*
	 * ut_user may not have the terminating NUL.
	 */
	strncpy (user, ut->ut_user, sizeof (ut->ut_user));
	user[sizeof (ut->ut_user)] = '\0';

	(void) time (&now);

	/*
	 * Check if they are allowed to be logged in right now.
	 */
	if (!isttytime (user, ut->ut_line, now)) {
		return 0;
	}
	return 1;
}


static void send_mesg_to_tty (int tty_fd)
{
	TERMIO oldt, newt;
	FILE *mesg_file, *tty_file;
	int c;
	bool is_tty;

	tty_file = fdopen (tty_fd, "w");
	if (NULL == tty_file) {
		return;
	}

	is_tty = (GTTY (tty_fd, &oldt) == 0);
	if (is_tty) {
		/* Suggested by Ivan Nejgebauar <ian@unsux.ns.ac.yu>:
		   set OPOST before writing the message. */
		newt = oldt;
		newt.c_oflag |= OPOST;
		STTY (tty_fd, &newt);
	}

	mesg_file = fopen (HUP_MESG_FILE, "r");
	if (NULL != mesg_file) {
		while ((c = getc (mesg_file)) != EOF) {
			if (c == '\n') {
				putc ('\r', tty_file);
			}
			putc (c, tty_file);
		}
		fclose (mesg_file);
	} else {
		fputs (DEFAULT_HUP_MESG, tty_file);
	}
	fflush (tty_file);
	fclose (tty_file);

	if (is_tty) {
		STTY (tty_fd, &oldt);
	}
}


/*
 * logoutd - logout daemon to enforce /etc/porttime file policy
 *
 *	logoutd is started at system boot time and enforces the login
 *	time and port restrictions specified in /etc/porttime. The
 *	utmpx/utmp file is periodically scanned and offending users are logged
 *	off from the system.
 */
int main (int argc, char **argv)
{
	int i;
	int status;
	pid_t pid;

#ifdef USE_UTMPX
	struct utmpx *ut;
#else				/* !USE_UTMPX */
	struct utmp *ut;
#endif				/* !USE_UTMPX */
	char user[sizeof (ut->ut_user) + 1];	/* terminating NUL */
	char tty_name[sizeof (ut->ut_line) + 6];	/* /dev/ + NUL */
	int tty_fd;

	if (1 != argc) {
		(void) fputs (_("Usage: logoutd\n"), stderr);
	}

	(void) setlocale (LC_ALL, "");
	(void) bindtextdomain (PACKAGE, LOCALEDIR);
	(void) textdomain (PACKAGE);

#ifndef DEBUG
	for (i = 0; close (i) == 0; i++);

	setpgrp ();

	/*
	 * Put this process in the background.
	 */
	pid = fork ();
	if (pid > 0) {
		/* parent */
		exit (EXIT_SUCCESS);
	} else if (pid < 0) {
		/* error */
		perror ("fork");
		exit (EXIT_FAILURE);
	}
#endif				/* !DEBUG */

	/*
	 * Start syslogging everything
	 */
	Prog = Basename (argv[0]);

	OPENLOG ("logoutd");

	/*
	 * Scan the utmpx/utmp file once per minute looking for users that
	 * are not supposed to still be logged in.
	 */
	while (true) {

		/* 
		 * Attempt to re-open the utmpx/utmp file. The file is only
		 * open while it is being used.
		 */
#ifdef USE_UTMPX
		setutxent ();
#else				/* !USE_UTMPX */
		setutent ();
#endif				/* !USE_UTMPX */

		/*
		 * Read all of the entries in the utmpx/utmp file. The entries
		 * for login sessions will be checked to see if the user
		 * is permitted to be signed on at this time.
		 */
#ifdef USE_UTMPX
		while ((ut = getutxent ()) != NULL)
#else				/* !USE_UTMPX */
		while ((ut = getutent ()) != NULL)
#endif				/* !USE_UTMPX */
		{
			if (ut->ut_type != USER_PROCESS) {
				continue;
			}
			if (ut->ut_user[0] == '\0') {
				continue;
			}
			if (check_login (ut)) {
				continue;
			}

			/*
			 * Put the rest of this in a child process. This
			 * keeps the scan from waiting on other ports to die.
			 */

			pid = fork ();
			if (pid > 0) {
				/* parent */
				continue;
			} else if (pid < 0) {
				/* failed - give up until the next scan */
				break;
			}
			/* child */

			if (strncmp (ut->ut_line, "/dev/", 5) != 0) {
				strcpy (tty_name, "/dev/");
			} else {
				tty_name[0] = '\0';
			}

			strcat (tty_name, ut->ut_line);
#ifndef O_NOCTTY
#define O_NOCTTY 0
#endif
			tty_fd =
			    open (tty_name, O_WRONLY | O_NDELAY | O_NOCTTY);
			if (tty_fd != -1) {
				send_mesg_to_tty (tty_fd);
				close (tty_fd);
				sleep (10);
			}

			if (ut->ut_pid > 1) {
				kill (-ut->ut_pid, SIGHUP);
				sleep (10);
				kill (-ut->ut_pid, SIGKILL);
			}

			strncpy (user, ut->ut_user, sizeof (user) - 1);
			user[sizeof (user) - 1] = '\0';

			SYSLOG ((LOG_NOTICE,
				 "logged off user '%s' on '%s'", user,
				 tty_name));

			/*
			 * This child has done all it can, drop dead.
			 */
			exit (EXIT_SUCCESS);
		}

#ifdef USE_UTMPX
		endutxent ();
#else				/* !USE_UTMPX */
		endutent ();
#endif				/* !USE_UTMPX */

#ifndef DEBUG
		sleep (60);
#endif
		/*
		 * Reap any dead babies ...
		 */
		while (wait (&status) != -1);
	}

	return EXIT_FAILURE;
}

